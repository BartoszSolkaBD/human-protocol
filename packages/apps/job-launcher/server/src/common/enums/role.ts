export enum Role {
  Exchange = 'exchange',
  Recording = 'recording',
  Reputation = 'reputation',
  JobLauncher = 'job-launcher',
}
