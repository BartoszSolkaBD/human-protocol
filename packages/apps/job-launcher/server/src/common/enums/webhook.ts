export enum EventType {
  ESCROW_CREATED = 'escrow_created',
  ESCROW_CANCELED = 'escrow_canceled',
  SUBMISSION_REJECTED = 'submission_rejected',
}

export enum OracleType {
  FORTUNE = 'fortune',
  CVAT = 'cvat',
  HCAPTCHA = 'hcaptcha',
}

export enum WebhookStatus {
  PENDING = 'PENDING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
}
