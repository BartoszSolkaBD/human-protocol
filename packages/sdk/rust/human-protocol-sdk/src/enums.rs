use chrono::{DateTime, Utc};
use thiserror::Error;
use serde_derive::{Deserialize, Serialize};

/// Represents various blockchain network IDs.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum ChainId {
    Mainnet = 1,
    Rinkeby = 4,
    Goerli = 5,
    BscMainnet = 56,
    BscTestnet = 97,
    Polygon = 137,
    PolygonMumbai = 80001,
    Moonbeam = 1284,
    MoonbaseAlpha = 1287,
    Avalanche = 43114,
    AvalancheTestnet = 43113,
    Skale = 1273227453,
    Localhost = 1338
}

/// Represents the status of an escrow.
#[derive(Debug, Clone, PartialEq)]
pub enum EscrowStatus {
    /// Escrow is launched.
    Launched,
    /// Escrow is funded, and waiting for the results to be submitted.
    Pending,
    /// Escrow is partially paid out.
    Partial,
    /// Escrow is fully paid.
    Paid,
    /// Escrow is finished.
    Complete,
    /// Escrow is cancelled.
    Cancelled,
}

impl EscrowStatus {
    pub fn to_string_vec() -> Vec<String> {
        vec![
            String::from("Launched"),
            String::from("Pending"),
            String::from("Partial"),
            String::from("Paid"),
            String::from("Complete"),
            String::from("Cancelled"),
        ]
    }
}

pub enum Role {
  JobLauncher,
  ExchangeOracle,
  RecordingOracle,
  ReputationOracle
}

impl Role {
    pub fn to_string_vec() -> Vec<String> {
        vec![
            String::from("Job Launcher"),
            String::from("Exchange Oracle"),
            String::from("Recording Oracle"),
            String::from("Reputation Oracle")
        ]
    }
}


#[derive(Debug, Error, PartialEq, Eq)]
pub enum EthereumError {
    #[error("Invalid Ethereum Address Error")]
    InvalidEthereumAddressError,
    #[error("Transaction Failed Error")]
    TransactionFailedError,
    #[error("Invalid Argument Error")]
    InvalidArgumentError,
    #[error("Web3 Error: {0}")]
    Web3Error(String),
    #[error("Web3 Contract Error: {0}")]
    Web3ContractError(String),
    #[error("Web3 Event Error: {0}")]
    Web3EventError(String),
    #[error("Event Parsing Error")]
    EventParsingError,
    #[error("Invalid Fee Error")]
    InvalidFeeError,
    #[error("Invalid Url Error")]
    InvalidUrlError,
    #[error("Invalid Key Error")]
    InvalidKeyError,
    #[error("Empty Hash Error")]
    EmptyHashError,
    #[error("Invalid Hash Error {0} != {1}")]
    InvalidHashError(String, String),
    #[error("Escrow Address Not Provided By Factory Error")]
    EscrowAddressNotProvidedByFactoryError,
    #[error("Unknown Escrow Status Error")]
    UnknownEscrowStatusError,
    #[error("Escrow Has Invalid Status Error")]
    EscrowHasInvalidStatusError,
    #[error("Invalid Recipient Size Error")]
    InvalidRecipientsSizeError,
    #[error("Invalid Amounts Size Error")]
    InvalidAmountsSizeError,
    #[error("Invalid Trusted Handlers Size Error")]
    InvalidTrustedHandlersSizeError,
    #[error("Recipients And Amounts Must Be Same Length Error")]
    RecipientsAndAmountsMustBeSameLengthError,
    #[error("Invalid Keys Size Error")]
    InvalidKeysSizeError,
    #[error("Invalid Values Size Error")]
    InvalidValuesSizeError,
    #[error("Keys And Values Must Be Same Length Error")]
    KeysAndValuesMustBeSameLengthError,
    #[error("Escrow Does Not Have Enough Balance Error")]
    EscrowDoesNotHaveEnoughBalanceError,
    #[error("Invalid ChainId Error")]
    InvalidChainIdError,
    #[error("Invalid Dates Error: {0} must be earlier than {1}")]
    InvalidDatesError(DateTime<Utc>, DateTime<Utc>),
    #[error("Network Not Found For Chain Id Error: {0}")]
    NetworkNotFoundForChainIdError(u32),
    #[error("Escrow Not Found For Escrow Address Error: {0}")]
    EscrowNotFoundForEscrowAddressError(String),
    #[error("No Logs Found Error")]
    NoLogsFoundError,
    #[error("Invalid Token Address Found Error")]
    InvalidTokenAddressFoundError,
    #[error("Log Parsing Error: {0}")]
    LogParsingError(String),
    #[error("Escrow Cancel Error")]
    EscrowCancelError,
    #[error("Unexpected Status Code Error {0}")]
    UnexpectedStatusCodeError(u16),
    #[error("Request Error {0}")]
    RequestError(String),
    #[error("Non Positive Value Error")]
    NonPositiveValueError,
    #[error("Rewards Not Found For Slasher Address Error: {0}")]
    RewardsNotFoundForSlasherAddressError(String),
    #[error("Leaders Not Found Error")]
    LeadersNotFoundError,
    #[error("Leader Not Found Error: {0}")]
    LeaderNotFoundError(String),
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum Error {
    #[error("Invalid Token Conversion Error")]
    InvalidTokenConversionError,
    #[error("Invalid Tuple Length Error")]
    InvalidTupleLengthError
    
}
