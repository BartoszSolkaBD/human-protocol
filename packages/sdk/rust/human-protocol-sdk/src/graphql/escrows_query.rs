use chrono::{DateTime, Utc};
use cynic::{impl_scalar};
use serde::{Serialize};


use crate::enums::EscrowStatus;

mod schema {
  cynic::use_schema!("schemas/escrow.graphql");
}

impl_scalar!(i64, schema::I64);

#[derive(Serialize, Debug, Clone, cynic::QueryFragment)]
#[cynic(schema_path = "schemas/escrow.graphql")]
pub struct EscrowData {
  pub id: String,
  pub address: String,
  pub amount_paid: cynic::Id,
  pub balance: cynic::Id,
  pub count: cynic::Id,
  pub factory_address: String,
  pub launcher: String,
  pub status: Option<String>,
  pub token: String,
  pub total_funded_amount: cynic::Id,
  pub created_at: String,
  pub final_results_url: Option<String>,
  pub intermediate_results_url: Option<String>,
  pub manifest_hash: Option<String>,
  pub manifest_url: Option<String>,
  pub recording_oracle: Option<String>,
  pub recording_oracle_fee: Option<cynic::Id>,
  pub reputation_oracle: Option<String>,
  pub reputation_oracle_fee: Option<cynic::Id>,
  pub exchange_oracle: Option<String>,
  pub exchange_oracle_fee: Option<cynic::Id>,
}

#[derive(cynic::QueryVariables, Clone)]
pub struct EscrowFilter {
  launcher: Option<String>,
  exchange_oracle: Option<String>,
  recording_oracle: Option<String>,
  reputation_oracle: Option<String>,
  job_requester_id: Option<String>,
  status: Option<Vec<String>>,
  from: Option<String>,
  to: Option<String>,
}


impl Default for EscrowFilter {
  fn default() -> Self { 
    let current_datetime: DateTime<Utc> = Utc::now();
    let timestamp: i64 = current_datetime.timestamp();

    Self {
      launcher: Some("0x".into()),
      exchange_oracle: None,
      recording_oracle: Some("0x".into()),
      reputation_oracle: Some("0x".into()),
      job_requester_id: None,
      status: Some(EscrowStatus::to_string_vec()),
      from: Some("0".to_string()),
      to: Some(timestamp.to_string()),
    }
  }
}

impl EscrowFilter {
  pub fn launcher(mut self, launcher: Option<String>) -> Self {
    self.launcher = launcher;
    self
  }

  pub fn exchange_oracle(mut self, exchange_oracle: Option<String>) -> Self {
    self.exchange_oracle = exchange_oracle;
    self
  }

  pub fn recording_oracle(mut self, recording_oracle: Option<String>) -> Self {
    self.recording_oracle = recording_oracle;
    self
  }

  pub fn reputation_oracle(mut self, reputation_oracle: Option<String>) -> Self {
    self.reputation_oracle = reputation_oracle;
    self
  }

  pub fn job_requester_id(mut self, job_requester_id: Option<String>) -> Self {
    self.job_requester_id = job_requester_id;
    self
  }

  pub fn status(mut self, status: Option<Vec<String>>) -> Self {
    self.status = status;
    self
  } 

  pub fn from(mut self, from: Option<String>) -> Self {
    self.from = from;
    self
  }

  pub fn to(mut self, to: Option<String>) -> Self {
    self.to = to;
    self
  }

  pub fn build(self) -> EscrowFilter {
      EscrowFilter {
          launcher: self.launcher,
          exchange_oracle: self.exchange_oracle,
          recording_oracle: self.recording_oracle,
          reputation_oracle: self.reputation_oracle,
          job_requester_id: self.job_requester_id,
          status: self.status,
          from: self.from,
          to: self.to,
      }
  }
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Root", variables = "EscrowFilter")]
#[cynic(schema_path = "schemas/escrow.graphql")]
pub struct GetEscrowsQuery {
  #[arguments(
    where: { 
      launcher_contains: $launcher,  
      exchangeOracle: $exchange_oracle,
      recordingOracle_contains: $recording_oracle,
      reputationOracle_contains: $reputation_oracle,
      jobRequesterId: $job_requester_id,
      status_in: $status,
      createdAt_gte: $from,
      createdAt_lte: $to
    }
  )]
  pub escrows: Option<Vec<EscrowData>>,
}

pub async fn run_query(url: &str, filter: &EscrowFilter) -> cynic::GraphQlResponse<GetEscrowsQuery> {
  use cynic::http::ReqwestExt;

  let query = build_query(filter);
  
  reqwest::Client::new()
    .post(url)
    .run_graphql(query)
    .await
    .unwrap()
}

fn build_query(filter: &EscrowFilter) -> cynic::Operation<GetEscrowsQuery, EscrowFilter> {
  use cynic::QueryBuilder;

  GetEscrowsQuery::build(filter.clone())
}

#[cfg(test)]
mod test {
    use crate::{enums::ChainId, constants::NETWORKS};

    use super::*;
    use tokio::test;

    #[test]
    async fn snapshot_test_query() {
      let filter = EscrowFilter::default();
      let query = build_query(&filter);

      insta::assert_snapshot!(query.query);
    }

    #[tokio::test]
    async fn test_running_query() {
      let network = NETWORKS.get(&ChainId::PolygonMumbai).unwrap();
      let filter = EscrowFilter::default();
      let result = run_query(network.subgraph_url, &filter).await;
      if let Some(errors) = &result.errors {
          assert_eq!(errors.len(), 0);
      }
      insta::assert_debug_snapshot!(result.data);
    }

    #[tokio::test]
    async fn test_filter_builder() {
        let filter = EscrowFilter::default()
            .launcher(Some("some_launcher".to_string()))
            .exchange_oracle(Some("some_exchange_oracle".to_string()))
            .recording_oracle(Some("some_recording_oracle".to_string()))
            .reputation_oracle(Some("some_reputation_oracle".to_string()))
            .job_requester_id(Some("some_job_requester_id".to_string()))
            .status(Some(vec!["status1".to_string(), "status2".to_string()]))
            .from(Some("start_date".to_string()))
            .to(Some("end_date".to_string()))
            .build();

        assert_eq!(filter.launcher, Some("some_launcher".to_string()));
        assert_eq!(
            filter.exchange_oracle,
            Some("some_exchange_oracle".to_string())
        );
        assert_eq!(
            filter.recording_oracle,
            Some("some_recording_oracle".to_string())
        );
        assert_eq!(
            filter.reputation_oracle,
            Some("some_reputation_oracle".to_string())
        );
        assert_eq!(
            filter.job_requester_id,
            Some("some_job_requester_id".to_string())
        );
        assert_eq!(
            filter.status,
            Some(vec!["status1".to_string(), "status2".to_string()])
        );
        assert_eq!(filter.from, Some("start_date".to_string()));
        assert_eq!(filter.to, Some("end_date".to_string()));
    }
}