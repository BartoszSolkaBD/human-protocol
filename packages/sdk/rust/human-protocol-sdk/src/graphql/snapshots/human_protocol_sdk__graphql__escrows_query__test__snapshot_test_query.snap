---
source: src/graphql/escrows_query.rs
expression: query.query
---
query GetEscrowsQuery($launcher: String, $exchangeOracle: String, $recordingOracle: String, $reputationOracle: String, $jobRequesterId: String, $status: [String!], $from: String, $to: String) {
  escrows(where: {launcher_contains: $launcher, exchangeOracle: $exchangeOracle, recordingOracle_contains: $recordingOracle, reputationOracle_contains: $reputationOracle, jobRequesterId: $jobRequesterId, status_in: $status, createdAt_gte: $from, createdAt_lte: $to, }) {
    id
    address
    amountPaid
    balance
    count
    factoryAddress
    launcher
    status
    token
    totalFundedAmount
    createdAt
    finalResultsUrl
    intermediateResultsUrl
    manifestHash
    manifestUrl
    recordingOracle
    recordingOracleFee
    reputationOracle
    reputationOracleFee
    exchangeOracle
    exchangeOracleFee
  }
}


